### ISSUE TITLE
Не меняется раскладка "pя3 "

### ISSUE THREAD

**Issue by 74ls00** [BladeMight/Mahou#214]
Я что не могу изменять раскладку части слова или слова с цифрой ?

Собственно волшебный текст, теги я так понял в лисе решили отключить типо он устарел.

pя3

**Comment by BladeMight** [comment]
> Я что не могу изменять раскладку части слова или слова с цифрой ?

@74ls00 Тут присутствуют символы из русской `я`, английской `p` и обоих `3`, при попытке конвертации Mahou не может определить в какой раскладке текст, т.к. равное количество символов из разных раскладок в режиме "Считать раскладку для всего слова", можно его выключить и без него будет _по-символьная конвертация на другую раскладку каждого символа_, пример: `pя3` -> `зя3` -> `pz3`.

**Comment by BladeMight** [comment]
Можно подумать над улучшением `Считать раскладку для всего слова`.

**Comment by 74ls00** [comment]
Я так понял проблема в том что опять  прога не успела переключить раскладку потомучто в браузере 300 вкладок и 16гб оперативы для этого маловато.
Попробую изменить приоритет jkl.

Так этож раскладка, пусть меняет для выделеного без определения, тогда бы я понял что слово в разных раскладках, потомучто  комп опять тормознул.
Она же както определяет знак по коду, вроде как  наши и их буквы хоть и похожи, но с разными кодами и прога смотрит именно код.

Похоже по невнимательности втупил и я и прога

**Comment by 74ls00** [comment]
Клавиши смены раскладки и запятая расположены близко и часто требуется переключить раскладку текста начиная с запятой из за случайного нажатия.
Запятая это знак вопроса.
При выделении фрагмента текста начиная с вопроса, раскладка переключается у всей строки английских знаков за исключением первого знака вопроса который тоже должен быть изменён на запятую.
Может вы сделаете просто переключение всех выделенных знаков без распознания ?

**Comment by BladeMight** [comment]
"Может вы сделаете просто переключение всех выделенных знаков без распознания ?"

Если выключить "Считать раскладку для всего слова" так и будет, но к сожалению не получится ничего хорошего, т.к. в первый раз конвертации будет `?` заменен на `,` во второй `б` и в третий опять на `,` и так по кругу.

> раскладка переключается у всей строки английских знаков за исключением первого знака

Вообще символы без "букв их языка" рядом невозможно распознать из какой раскладки они были напечатаны и в итоге дает то что символ просто проигнорировался. Можно попробовать добавить функцию как "Считать раскладку для всего слова", только для всего выделенного фрагмента, т.е. `Считать раскладку для всех слов в выделении.`, т.е. будет определятся раскладка всех слов, а потом тех что больше будет считаться раскладкой выделения и по ней будет производится коновертация всего выделенного, включая "символов одиночек".

**Comment by 74ls00** [comment]
Было бы не плохо менять раскладку для всего выделенного без распознания.
Но чтото  я ничего включить не могу.

**Comment by 74ls00** [comment]
Было бы не плохо менять раскладку для всего выделенного без распознания.
Но чтото  я ничего включить не могу.

![default](https://user-images.githubusercontent.com/11761076/50382974-cb3e0080-06bb-11e9-8dc4-4618001ac75b.png)
Если убрать птичку "считывать для всего слова", верхний пункт становится красным и ничего не сохраняется
![_cr](https://user-images.githubusercontent.com/11761076/50382981-0809f780-06bc-11e9-82ee-48470c6ef107.png)

**Comment by BladeMight** [comment]
> После применения становится как на первом скрине

Функция "Смена раскладки в Конверт выделения" без функции "Считать раскладку для всего слова в КВ", - это не просто смена раскладки - это совершенно другой тип распознавания + смена раскладки, использующий текущую(раскладку окна, так что если в консольном окне, то без JKL или getconkbl.dll работать не будет) и следующую(в зависимости от того как стоит переключение, между всеми или только между 2-мя во вкладке Раскладки) раскладку, т.е. считает что все выделенные символы написаны в раскладке которая сейчас выбрана. Об остальных функциях почитай подсказки. И красные они потому что уже не такие стабильные и не обновляются, с недавнего времени, в пользу `Считать раскладку для всего слова в КВ`, но все еще оставлены и работают для тех кому это нужно.

В общем то что тебе надо - это как раз только "Смена раскладки в Конверт выделения", как здесь:

![](https://user-images.githubusercontent.com/11761076/50382990-47d0df00-06bc-11e9-9b32-2e726a4469d5.png)
 
Попробуй теперь выделяя твое прошлое от предложение от знака `?`, 
```
Дело было так, переключил раскладку.
```

Получаем:

1. Если раскладка до конвертации была русской:
```
Дело было так& gthtrk/xbk hfcrkflre|
```
2. Если раскладка до конвертации была английской:
```
Дело было так, переключил раскладку.
```
Видишь? №2 - как раз то что тебе нужно.

**Comment by 74ls00** [comment]
Ладно буду выяснять почему не могу это включить. В релизе настройка сохраняется, но я там даже хоткеи не помню как настраивать.
Видимо какойто конфликт или баг

**Comment by 74ls00** [comment]
Это вообще всё не то, как раз создаёт первую проблему.
Если включен "конверт выделения", то язык переключается вместе с текстом, а мне нужно чтоб он менялся только по хоткею и никогда сам.

Если просто не включать "для всего слова", то вообще выходит тарабарщина, конвертируются некоторые знаки причём только в одну раскладку, и то через раз.

Если включить "конверт+" ох не зря там красным написано. Один раз работает, потом начисто дохнут все хоткеи, прокрутка в системе и вообще будто падает проводник, но только даже во всех програмах, помогает только перезагрузка системы, этот пункт вообще лучше никогда не включать.

Наверно стоит пересмотреть эти 3 пункта настройки и разделить. В них явно используется по несколько разных функций не совпадающих ниразу с другими пунктами, вот эти функции и надо разделить.

**Comment by BladeMight** [comment]
> Наверно стоит переделать эти 3 пункта

Да, я думал на тем чтобы сделать возможность выбирать методику конвертации и дополнительные фичи вроде "смены раскладки после конвертации". И к тому же эти "методики" требуют полного обновления, т.к. многие из них не работают как нужно. Вообще если говорим о "двунаправленной конвертации" и "конвертации слов содержащих символы/буквы из обоих раскладок" - то тут не все так просто, раннее была тема: #18, там я приделал "конверт+", который абы-как но работал, но с момента появления JKL и других функций, работоспособность и совместимость между ними имела много нехороших последствий. 
Думаю на это уйдет какое-то время, ну и понимая что сделать придется многое нужно еще и решится начать... Так что пока будет так, может если скажешь как конкретно ты хочешь(подробно опиши с множеством примеров, с символами, с буквами из обоих раскладок в одном слове, поддержкой более 2 раскладок и т.п.) то может специально для тебя я быстро смогу сделать "особую конвертацию".
